docs/ref-manual/cvl/methods.md

```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
+
+Whether a function call is replaced by an approximation depends on the context
+in which the function is called in addition to the application policy for its
+signature.  If present, the application policy must be either `ALL` or
+`UNRESOLVED`; the default policy is TODO.  The decision is made as follows:
+
+ * If the function is called from CVL rather than from contract code then it is
+   never replaced by a summary.
+
+ * If the code for the function is known at verification time, either because
+   it is a method of `currentContract` or because the receiver contract is
+   {ref}`linked <linking>`, then the function is only summarized if the
+   resolution type is `ALL`.
 
-### Application policies (UNRESOLVED or ALL)
+ * If the code for the function is not known at verification time, then the
+   function call must be summarized.  If no summary is given, the default summary
+   type is {ref}`AUTO <auto-summary>`, whose behavior is determined by the type of
+   function call.  In this case, the verification report will contain a contract
+   call resolution warning.
 
-### `ALWAYS`, `CONSTANT`, `PER_CALLEE_CONSTANT`
+```{todo}
+The default application policy is currently undocumented.
+```
+
+```{todo}
+The old documentation is ambiguous about the behavior of `UNRESOLVED` summaries
+for internal methods.
+```
+
+Method summaries apply to all calls, regardless of the receiver address.  There
+is currently no way to apply different summaries to different contracts or to
+summarize some calls and not others to methods with the same ABI signature.
+For this reason, it is not possible to specify a summary for a method that is
+qualified by a contract name.
+
+Summary types
+-------------
+
+(const-summary)=
+### View summaries: `ALWAYS`, `CONSTANT`, `PER_CALLEE_CONSTANT`, and `NONDET`
+
+These four summary types treat the summarized methods as view methods: the
+summarized methods are replaced by approximations that do not update the state
+of any contract.  They differ in the assumptions made about the return value:
+
+ * The `ALWAYS(v)` approximation assumes that the method always returns `v`
+
+ * The `CONSTANT` approximation assumes that all calls to methods with the given
+   signature always return the same result.
+
+ * The `PER_CALLEE_CONSTANT` approximation assumes that all calls to the method
+   on a given receiver contract must return the same result, but that the
+   returned value may be different for different receiver contracts.
+
+ * The `NONDET` approximation makes no assumptions about the return values; each
+   call to the summarized method may return a different result.
+
+```{todo}
+The old documentation seems to imply that the ETH balances for calls to
+`ALWAYS`, `CONSTANT`, and `PER_CALLEE_CONSTANT` do not account for the ETH value
+passed in the function call, while `NONDET` does.
+
+This seems like the wrong behavior?
+```
+
+```{todo}
+The following note from the old documentation needs clarification:
```

Comment:
@jtoman ?

Reply:
Yeah, this is pretty in the weeds, but the amount of data returned by the caller does not necessarily need to honor the request of the caller (in fact, in the case of dynamically sized return payloads, there is no "up front" size that the caller can declare for the expected size of the return buffer).

Instead, there is a sequence of sanity checks in the caller to establish that the callee returned enough data to satisfy the expected return values. So, if contract A calls contract B expecting contract B to return two uint256 values, and contract B only returns 32 bytes, then the call within A fails.

In other words, too-small return buffers is a failure mode of calls that are completely ignored by the summaries: from the perspective of the caller, the call external is summarized in such a way that the call "appears" to have exactly returned the amount of data required.

The optimistic return size is a flag to avoid insufficient returndata reverts (presumably when running with `@withrevert`). Like the summaries above, with the optimisticReturnsize flag, any unresolved (havocing) function will still be assumed to return the necessary number of bytes (even if those bytes are completely arbitrary).

Reply:
I think I have accurately described the situation.